### DISCUSSION TITLE
pid and fd & backlog for SYN packets

### DISCUSSION THREAD

**Discussion by seohyeon0322** [ANLAB-KAIST/KENSv3#58] [Project Q&A 2. Transport Layer (TCP)]
1. Can different process(pid) have the same file descriptor(fd)? Or file descriptor is an unique value?

2. I am confused about "backlog"in listen(). I thought the "backlog" limits the amount of SYN packet server can get. However, in TestAccept_Backlog2, backlog is 1 but expected_accept is 6. I thought expected_accept should be 1 as backlog is 1. Is my thought  wrong?

**Comment by Pusnow** [comment]
> Can different process(pid) have the same file descriptor(fd)? Or file descriptor is an unique value?

Each process has its own fd space, and different process can have the same fd number (number only, they point different connections).

> I am confused about "backlog"in listen(). I thought the "backlog" limits the amount of SYN packet server can get. However, in TestAccept_Backlog2, backlog is 1 but expected_accept is 6. I thought expected_accept should be 1 as backlog is 1. Is my thought wrong?

It is a little bit tricky. In KENS, the `backlog` parameter limits the number of pre-established (currently handshaking) connections, not accepted connections.
Even if the listing peer does not call `accept()`, the connections can be established. The `accept()` only pops a connection from list of established connections.

https://github.com/ANLAB-KAIST/KENSv3/blob/a61076151479c20c8e26ca259f4d50e9eb92e4af/app/kens/testhandshake.cpp#L288-L313

In `Backlog2`, the `connect()` intervals are about 10 microseconds, which is enough time for handshaking. Thus all connections can be accepted.



### If you are interested more, read below.

The effect of the `backlog` parameter is implementation-specific, and there are differences among implementations.
In modern Linux, the `backlog` parameter means the queue length for completely established sockets waiting to be accepted.
Linux before v2.2, the `backlog` parameter means the number of incomplete connection requests (same as KENS).

See NOTES in `listen(2)`:

https://github.com/ANLAB-KAIST/KENSv3/blob/22b8748332f44905abb3277dad14fcc088bc36c6/Documentation/man/listen.2.txt#L60-L64

FreeBSD also has a different meaning (https://www.freebsd.org/cgi/man.cgi?query=listen&sektion=2&format=html).
